bindings/src/main/java/io/pravega/storage/hdfs/HDFSChunkStorage.java

```diff
@@ -0,0 +1,473 @@
+import io.pravega.segmentstore.storage.chunklayer.ChunkHandle;
+import io.pravega.segmentstore.storage.chunklayer.ChunkInfo;
+import io.pravega.segmentstore.storage.chunklayer.ChunkNotFoundException;
+import io.pravega.segmentstore.storage.chunklayer.ChunkStorage;
+import io.pravega.segmentstore.storage.chunklayer.ChunkStorageException;
+import io.pravega.segmentstore.storage.chunklayer.ConcatArgument;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsAction;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.io.IOUtils;
+
+/***
+ *  {@link ChunkStorage} for HDFS based storage.
+ *
+ * Each chunk is represented as a single Object on the underlying storage.
+ *
+ * This implementation works under the assumption that data is only appended and never modified.
+ * The concat operation is implemented using HDFS native concat operation.
+ */
+
+@Slf4j
+class HDFSChunkStorage extends BaseChunkStorage {
+    private static final FsPermission READWRITE_PERMISSION = new FsPermission(FsAction.READ_WRITE, FsAction.NONE, FsAction.NONE);
+    private static final FsPermission READONLY_PERMISSION = new FsPermission(FsAction.READ, FsAction.READ, FsAction.READ);
+
+    //region Members
+
+    private final HDFSStorageConfig config;
+    private FileSystem fileSystem;
+    private final AtomicBoolean closed;
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the HDFSStorage class.
+     *
+     * @param config The configuration to use.
+     */
+    HDFSChunkStorage(HDFSStorageConfig config) {
+        Preconditions.checkNotNull(config, "config");
+        this.config = config;
+        this.closed = new AtomicBoolean(false);
+        initialize();
+    }
+
+    //endregion
+
+    //region capabilities
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports merge operation either natively or through appends.
+     *
+     * @return True or false.
+     */
+    @Override
+    public boolean supportsConcat() {
+        return true;
+    }
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports append operation on underlying storage object.
+     *
+     * @return True or false.
+     */
+    @Override
+    public boolean supportsAppend() {
+        return true;
+    }
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports truncate operation on underlying storage object.
+     *
+     * @return True or false.
+     */
+    @Override
+    public boolean supportsTruncation() {
+        return false;
```

Comment:
I also thought that HDFS had a truncate operation, is there a reason for us to say that it doesn't support truncate?

https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/FileSystemShell.html#truncate

Reply:
We need supportsTruncation to support truncation at front, not end.

Reply:
I think one way to disambiguate this is to name this operation `HeadTruncation`. We may want to rename these methods to `supportsHeadTruncation`. We can do it in a future issue.